### PR TITLE
Plugins: Improve performance of plugin selectors

### PR DESCRIPTION
**Pull request by robertf4**
#### Proposed Changes

This fixes some performance issues with the plugins selectors by reducing runtime complexity, indexing, and memoizing where possible.

#### Testing Instructions

1. Visit `/plugins/manage` in your dev environment and compare it to what you see in production. There should be no difference.
2. Visit /plugins/manage as an agency with many sites. See this Slack comment for an example: p1669769472607889-slack-C039K7CD119. Verify that the client is responsive after the initial network calls are completed. If the agency has more than 20 plugins then note that the icons and titles from the plugin repository will not show.
3. Do regression testing on the other plugins pages: `/plugins/manage/:site` and `/plugins/:plugin_slug`

#### Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Related to #

### CODE REVIEW

File: client/state/plugins/installed/selectors.js
Comment (robertf4) [inline client/state/plugins/installed/selectors.js:149]: `getPlugins` was a deceptively simple name compared to all the computation that was going on here and was arguably masking the very computationally expensive things being done previously, so I renamed it to `getFilteredAndSortedPlugins` to make it clear that there is more going on than simply getting plugins.

File: client/state/plugins/installed/selectors.js
Comment (robertf4) [inline client/state/plugins/installed/selectors.js:81]: This selector serves to perform the computation previously being done in `getPlugins` once, memoize it, and expose it in a way that it can be accessed in O(1) time if you have a plugin slug.

File: client/state/plugins/installed/selectors.js
Comment (robertf4) [inline client/state/plugins/installed/selectors.js:123]: This selector restructures the data calculated in getAllPluginsIndexedByPluginSlug for O(1) access if you have a site ID, and memoizes that.

File: client/state/plugins/installed/selectors.js
Comment (robertf4) [inline client/state/plugins/installed/selectors.js:137]: Calling `getPlugins` in all of these remaining selectors was very computationally expensive even after I reworked that selector since these are often called with different sets of siteIds thus causing the selector to recalculate numerous times, and somehow it was causing a loop where a component was rendering different sets of siteIds infinitely and thus infinitely recalculating the selector.

I changed all these selectors to use the more stable/simple data provided in `getAllPluginsIndexedBySiteId` and `getAllPluginsIndexedByPluginSlug` which greatly improve the performance of all of them and fixed the rendering loop.

**Comment by robertf4** [comment]
I plan to do a pass on the unit tests, adding more cases and testing the new selectors. I just wanted to go ahead and get a review on the existing work to keep things moving on this PR.

File: client/state/plugins/installed/test/selectors.js
Comment (robertf4) [inline client/state/plugins/installed/test/selectors.js:24]: The site IDs in production are numbers and not strings so this was causing some tests to fail due to type mismatches. Making these IDs numbers fixed that.

File: client/state/plugins/installed/selectors.js
Comment (yashwin) [inline client/state/plugins/installed/selectors.js:159]: I feel combining the map functions into one will slightly speed up computation.

`.map( ( siteId ) => allPluginsIndexedBySiteId[ Number(siteId) ] )`

File: client/state/plugins/installed/selectors.js
Comment (yashwin) [inline client/state/plugins/installed/selectors.js:135]: I feel like we can move the map inside forEach which could speed up the process slightly.

**Review by yashwin — DISMISSED** [review]
Great job on this, @robertf4. The dashboard loads fine and was able to perform basic operations like search, switch tabs, etc., for this one - p1669769472607889-slack-C039K7CD119 even though I faced a slight delay(millisecond). The performance has increased significantly. I tested most of the operations with less number of plugins, and that seems to work fine, and I couldn't find any regression in Jetpack Cloud. I haven't tested the changes in Calypso Blue as I feel there is no difference in functionality. I have left a couple of comments. I suggest waiting for @vitozev's approval too, just to be sure before merging this.

**Review by vitozev — CHANGES_REQUESTED** [review]
Overall this PR works great, but I found one issue.

image

If you open the http://jetpack.cloud.localhost:3001/plugins/akismet (i.e. single plugin view), there are a few issues:
- The sites where plugin is marked as active are different between development environment and production.
- If you use the to enable/disable the plugin for a specific site, it will apply the same action to all other available sites.

Are you able to reproduce it?

**Comment by robertf4** [comment]
> The sites where plugin is marked as active are different between development environment and production.
> If you use the to enable/disable the plugin for a specific site, it will apply the same action to all other available sites.

Good catch. These were both actually display issues due to the `active` property being overwritten on the plugin object each time a specific site was added to it, so the plugin would display with the status of the last site in the list. I changed the data structure so that site specific properties are only found on the site objects unless the plugin is requested for a single site which will hopefully prevent this from happening again.

**Comment by vitozev** [comment]
Thanks for fixing this, Robert! It works, however, I think this PR causes some other inconsistencies.

If you try to update a specific plugin from the same view, it will always say that it is already up to date. IIRC, we checked some properties of the plugin object and dispatched this action so that it might be a similar issue.

Here's a quick video showing the problem:

Screen Capture on 2022-12-20 at 15-22-28

You can easily test this by downloading an older version of a specific plugin from [the advanced view](https://wordpress.org/plugins/jetpack/advanced/) and installing it on your test site.

**Comment by robertf4** [comment]
> If you try to update a specific plugin from the same view, it will always say that it is already up to date.

This one is fixed now.

**Review by vitozev — CHANGES_REQUESTED** [review]
I carefully tested all possible flows (e.g., single actions, batch actions, etc.) in the single-site plugin management view and multi-site plugin management view, and here's what I found:

1. In http://jetpack.cloud.localhost:3001/plugins/manage (can be reproduced also in `/plugins/manage/:site`), the `Deactivate` bulk action (_note that `Activate` is working as expected)_ isn't working as expected, it says that plugins were deactivated, but actually, they are not (no API request is triggered).  
3. In http://jetpack.cloud.localhost:3001/plugins/manage (can be reproduced also in `/plugins/manage/:site`), the `Disable` bulk action (_note that `Enable` is working as expected_) for autoupdate isn't working as expected, the same problem as above.
4. In http://jetpack.cloud.localhost:3001/plugins/akismet, if you click `Update plugin`, it always says "Plugin already up to date" without making an API request. I think you already fixed a similar issue.

The easiest way to test these would be to load production in one tab, and the development environment in another, and you will spot the differences immediately.

**Comment by robertf4** [comment]
I've addressed some of the issues, but I'm still investigating more. Currently I'm seeing a site be marked as autoupdate disabled in my dev environment but enabled in prod, and it seems to be coming from the server this way.

**Comment by vitozev** [comment]
@robertf4, let me know if this PR can be reviewed again.

**Comment by robertf4** [comment]
Closed in favor of using https://github.com/Automattic/wp-calypso/pull/72363

**Pull request closed without merging**
